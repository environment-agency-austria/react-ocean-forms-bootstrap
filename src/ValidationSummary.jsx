/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Alert } from 'reactstrap';
import FontAwesomeIcon from '@fortawesome/react-fontawesome';
import faExclamationCircle from '@fortawesome/fontawesome-free-solid/faExclamationCircle';
import { ValidationSummary } from 'react-ocean-forms';

/**
 * Component for displaying a summary of all
 * validation errors of the form this component
 * lives in.
 */
function BootstrapValidationSummary(props) {
  const {
    title,
    ...rest
  } = props;

  return (
    <ValidationSummary
      title={title}
      {...rest}
      renderFieldError={(id, fieldName, errors, linkCallback) => (
        <li key={id}>
          <FontAwesomeIcon icon={faExclamationCircle} className="mr-2" />
          <a href="#" onClick={linkCallback}>
            {fieldName}: {errors}
          </a>
        </li>
      )}

      render={children => (
        <Alert color="danger" className="validation-summary">
          {children}
        </Alert>
      )}
    />
  );
}

BootstrapValidationSummary.defaultProps = {
  title: 'ojs_form_validationSummaryHeader',
};

BootstrapValidationSummary.propTypes = {
  title: PropTypes.string,
};
export default BootstrapValidationSummary;
