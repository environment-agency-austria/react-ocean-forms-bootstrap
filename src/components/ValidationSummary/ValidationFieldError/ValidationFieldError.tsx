/**
 * Copyright (c) 2019-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as React from 'react';

import { faExclamationCircle } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import { IValidationFieldErrorProps } from './ValidationFieldError.types';

export const ValidationFieldError: React.FC<IValidationFieldErrorProps> = (props) => {
  const { id, linkCallback, fieldName, errors } = props;

  return (
    <li key={id}>
      <FontAwesomeIcon icon={faExclamationCircle} className="mr-2" />
      <a href={`#${id}`} onClick={linkCallback}>
        {fieldName}
        :
        {' '}
        {errors}
      </a>
    </li>
  );
};
