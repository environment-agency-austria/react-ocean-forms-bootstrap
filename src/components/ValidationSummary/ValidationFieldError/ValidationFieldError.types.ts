import * as React from 'react';

export interface IValidationFieldErrorProps {
  id: string;
  fieldName: string;
  errors: React.ReactNode;
  linkCallback: React.MouseEventHandler;
}
