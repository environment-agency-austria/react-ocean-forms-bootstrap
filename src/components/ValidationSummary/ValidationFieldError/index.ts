export * from './ValidationFieldError';
export * from './ValidationFieldError.types';
