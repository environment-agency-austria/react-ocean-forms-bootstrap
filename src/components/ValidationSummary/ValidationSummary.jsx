/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Alert } from 'reactstrap';
import FontAwesomeIcon from '@fortawesome/react-fontawesome';
import faExclamationCircle from '@fortawesome/fontawesome-free-solid/faExclamationCircle';
import { ValidationSummary as CoreValidationSummary } from 'react-ocean-forms';

/**
 * Component for displaying a summary of all
 * validation errors of the form this component
 * lives in.
 */
function ValidationSummary(props) {
  const {
    title,
    ...rest
  } = props;

  return (
    <CoreValidationSummary
      title={title}
      {...rest}
      renderFieldError={(id, fieldName, errors, linkCallback) => (
        <li key={id}>
          <FontAwesomeIcon icon={faExclamationCircle} className="mr-2" />
          <a href="#" onClick={linkCallback}>
            {fieldName}
            :
            {' '}
            {errors}
          </a>
        </li>
      )}

      render={children => (
        <Alert color="danger" className="validation-summary">
          {children}
        </Alert>
      )}
    />
  );
}

ValidationSummary.displayName = 'ValidationSummary';

ValidationSummary.defaultProps = {
  title: 'ojs_form_validationSummaryHeader',
};

ValidationSummary.propTypes = {
  title: PropTypes.string,
};
export default ValidationSummary;
