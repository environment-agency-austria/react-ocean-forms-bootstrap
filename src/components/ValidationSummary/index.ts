export * from './ValidationSummary';
export * from './ValidationSummary.types';
export * from './ValidationFieldError';
