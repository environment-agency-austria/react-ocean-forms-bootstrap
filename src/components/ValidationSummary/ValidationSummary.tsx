/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as React from 'react';

import { ValidationSummary as CoreValidationSummary } from 'react-ocean-forms';
import { Alert } from 'reactstrap';

import { IValidationSummaryProps } from './ValidationSummary.types';
import { ValidationFieldError } from './ValidationFieldError';

/**
 * Component for displaying a summary of all
 * validation errors of the form this component
 * lives in.
 */
export class ValidationSummary extends React.Component<IValidationSummaryProps> {
  public static displayName: string = 'ValidationSummary';

  private renderFieldError = (
    id: string,
    fieldName: string,
    errors: React.ReactNode,
    linkCallback: React.MouseEventHandler,
  ): JSX.Element => {
    const { fieldErrorComponent: FieldErrorComponent = ValidationFieldError } = this.props;

    return (
      <FieldErrorComponent
        id={id}
        fieldName={fieldName}
        errors={errors}
        linkCallback={linkCallback}
      />
    );
  }

  private renderSummary = (children: JSX.Element): JSX.Element => {
    return (
      <Alert color="danger" className="validation-summary">
        {children}
      </Alert>
    );
  }

  public render(): JSX.Element {
    return (
      <CoreValidationSummary
        {...this.props}
        renderFieldError={this.renderFieldError}
        render={this.renderSummary}
      />
    );
  }
}
