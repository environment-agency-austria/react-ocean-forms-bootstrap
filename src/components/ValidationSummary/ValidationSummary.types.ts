import * as React from 'react';
import { PropsOf, Subtract, ValidationSummary } from 'react-ocean-forms';

import { IValidationFieldErrorProps } from './ValidationFieldError';

type TUpstreamValidationSummaryProps = PropsOf<typeof ValidationSummary>;

/**
 * Props for the ValidationSummary component
 */
export interface IValidationSummaryProps extends Subtract<TUpstreamValidationSummaryProps, { render: string; renderFieldError: string }> {
  fieldErrorComponent?: React.ComponentType<IValidationFieldErrorProps>;
}
