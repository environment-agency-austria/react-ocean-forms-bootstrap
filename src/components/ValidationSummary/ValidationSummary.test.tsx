import * as  React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';
import { ValidationSummary as CoreValidationSummary } from 'react-ocean-forms';

import { ValidationSummary } from './ValidationSummary';
import { IValidationSummaryProps } from './ValidationSummary.types';
import { IValidationFieldErrorProps } from './ValidationFieldError';

describe('<ValidationSummary />', () => {
  interface ISetupArgs {
    props?: Partial<IValidationSummaryProps>;
  }

  interface ISetupResult {
    wrapper: ShallowWrapper;
  }

  const setup = ({
    props,
  }: ISetupArgs = {}): ISetupResult => {
    const wrapper = shallow((
      <ValidationSummary id="mock-summary" {...props} />
    ));

    return {
      wrapper,
    };
  };

  it('should render without crashing', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  describe('render prop', () => {
    const { wrapper } = setup();
    const renderProp = wrapper.find(CoreValidationSummary).prop<((children: JSX.Element) => JSX.Element)>('render');

    it('should set the render prop correctly', () => {
      expect(typeof renderProp).toBe('function');
    });

    it('should correctly render the render prop', () => {
      const mockChildren = <div id="mock" />;
      const renderWrapper = shallow(renderProp(mockChildren));
      expect(renderWrapper).toMatchSnapshot();
    });
  });

  describe('renderFieldError prop', () => {
    const { wrapper } = setup();
    const renderFieldErrorProp = wrapper.find(CoreValidationSummary).prop<((...rest: unknown[]) => JSX.Element)>('renderFieldError');

    it('should set the render prop correctly', () => {
      expect(typeof renderFieldErrorProp).toBe('function');
    });

    it('should correctly render the render prop', () => {
      const renderWrapper = shallow(renderFieldErrorProp(
        'mock-id',
        'mock-name',
        'mock-error',
        jest.fn(),
      ));
      expect(renderWrapper).toMatchSnapshot();
    });
  });

  describe('fieldErrorComponent prop', () => {
    const CustomErrorRenderer: React.FC<IValidationFieldErrorProps> = () => {
      return <div id="mock-error" />;
    };

    const { wrapper } = setup({
      props: {
        fieldErrorComponent: CustomErrorRenderer,
      }
    });
    const renderFieldErrorProp = wrapper.find(CoreValidationSummary).prop<((...rest: unknown[]) => JSX.Element)>('renderFieldError');

    it('should accept custom field error components', () => {
      const renderWrapper = shallow(renderFieldErrorProp(
        'mock-id',
        'mock-name',
        'mock-error',
        jest.fn(),
      ));
      expect(renderWrapper).toMatchSnapshot();
    });
  });
});
