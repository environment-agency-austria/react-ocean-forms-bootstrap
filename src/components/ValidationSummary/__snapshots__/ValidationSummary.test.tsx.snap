// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ValidationSummary /> fieldErrorComponent prop should accept custom field error components 1`] = `
<div
  id="mock-error"
/>
`;

exports[`<ValidationSummary /> render prop should correctly render the render prop 1`] = `
<Fade
  appear={true}
  baseClass="fade"
  baseClassActive="show"
  className="validation-summary alert alert-danger"
  enter={true}
  exit={true}
  in={true}
  mountOnEnter={false}
  onEnter={[Function]}
  onEntered={[Function]}
  onEntering={[Function]}
  onExit={[Function]}
  onExited={[Function]}
  onExiting={[Function]}
  role="alert"
  tag="div"
  timeout={150}
  unmountOnExit={true}
>
  <div
    id="mock"
  />
</Fade>
`;

exports[`<ValidationSummary /> renderFieldError prop should correctly render the render prop 1`] = `
<li
  key="mock-id"
>
  <FontAwesomeIcon
    border={false}
    className="mr-2"
    fixedWidth={false}
    flip={null}
    icon={
      Object {
        "icon": Array [
          512,
          512,
          Array [],
          "f06a",
          "M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z",
        ],
        "iconName": "exclamation-circle",
        "prefix": "fas",
      }
    }
    inverse={false}
    listItem={false}
    mask={null}
    pull={null}
    pulse={false}
    rotation={null}
    size={null}
    spin={false}
    swapOpacity={false}
    symbol={false}
    title=""
    transform={null}
  />
  <a
    href="#mock-id"
    onClick={[MockFunction]}
  >
    mock-name
    :
     
    mock-error
  </a>
</li>
`;

exports[`<ValidationSummary /> should render without crashing 1`] = `
<FormComponent(ValidationSummary)
  id="mock-summary"
  render={[Function]}
  renderFieldError={[Function]}
/>
`;
